### PR TITLE
backupccl: create backup compaction iterator

### PR DESCRIPTION
**Pull request by kev-cao**
For the purposes of backup compaction, a custom iterator is required that behaves similarly to the ReadAsOfIterator, but also surfaces live tombstone point keys.

Epic: none

Release note: None

### CODE REVIEW

File: pkg/storage/backup_compaction_iterator.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+func (f *BackupCompactionIterator) NextKey() {
+	f.iter.NextKey()
+	f.advance()
+}
+
+func (f *BackupCompactionIterator) SeekGE(originalKey MVCCKey) {
+	synthetic := MVCCKey{Key: originalKey.Key, Timestamp: f.asOf}
+	f.iter.SeekGE(synthetic)
+	if f.advance(); f.valid && f.UnsafeKey().Less(originalKey) {
+		f.NextKey()
```
Comment (msbutler) [inline pkg/storage/backup_compaction_iterator.go:64]: leave a comment along the lines of: "see readAsOfIterator comment for explanation of this"

File: pkg/storage/backup_compaction_iterator_test.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+		t.Run(name, func(t *testing.T) {
+			batch := pebble.NewBatch()
+			defer batch.Close()
+			populateBatch(t, batch, test.input)
+			iter, err := batch.NewMVCCIterator(context.Background(), MVCCKeyAndIntentsIterKind, IterOptions{UpperBound: roachpb.KeyMax})
+			require.NoError(t, err)
+			defer iter.Close()
+
+			subtests := []iterSubtest{
+				{"NextKey", test.expectedNextKey, SimpleMVCCIterator.NextKey},
```
Comment (msbutler) [inline pkg/storage/backup_compaction_iterator_test.go:81]: could you add a `Next` subtest? It should return the same results.

File: pkg/storage/backup_compaction_iterator_test.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+		// Double skip across keys.
+		{input: "b2c2c1", expectedNextKey: "c1", asOf: "1"},
+
+		// Ensure next key captures at most one mvcc key per key after an asOf skip.
+		{input: "b3c2c1", expectedNextKey: "c2", asOf: "2"},
+
+		// Ensure an AOST 'next' takes precedence over a tombstone 'nextkey'.
+		{input: "b2Xb1c1", expectedNextKey: "b2Xc1", asOf: ""},
+		{input: "b2Xb1c1", expectedNextKey: "b1c1", asOf: "1"},
```
Comment (msbutler) [inline pkg/storage/backup_compaction_iterator_test.go:56]: add one more case of asOf 2? i expect b2xc1

File: pkg/storage/backup_compaction_iterator_test.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+		// Ensure the asOf timestamp moves the iterator during a seek.
+		{"a2a1", "a2", "a1", "1"},
+		{"a2b1", "a2", "b1", "1"},
+
+		// Ensure seek will return on a tombstone if it is the latest key.
+		{"a3Xa1b1", "a3", "a3X", ""},
+		{"a3Xa1c2Xc1", "b1", "c2X", ""},
+
+		// Ensure seek does not return on a key shadowed by a tombstone.
```
Comment (msbutler) [inline pkg/storage/backup_compaction_iterator_test.go:131]: this seek behavior is unrelated to the tombstone, we always seek to b1 because a2 and a1 are not the latest key given the aost. i'd rephrase the comment here.

File: pkg/storage/backup_compaction_iterator_test.go [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+		{"a3Xa2a1b1", "a2", "b1", "3"},
+		{"a3a2Xa1b1", "a1", "b1", ""},
+		{"a3a2Xa1b2Xb1c1", "a1", "b2X", ""},
+
+		// Ensure we can seek to a key right before a tombstone.
+		{"a2Xa1b2b1Xc1", "a1", "b2", ""},
+
+		// Ensure seeking on a tombstone still lands on the tombstone if
+		// within AOST
+		{"a2Xa1b2", "a2X", "a2X", ""},
```
Comment (msbutler) [inline pkg/storage/backup_compaction_iterator_test.go:142]: a tombstone is a value. we seek on mvcc keys (key + timestamp), not key values. A seek key should never have a tombstone. i think this is a good test case if you remove the "X" from the seek key and rephrase the comment.

File: pkg/storage/backup_compaction_iterator_test.go [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+		{"a2a1", "a2", "a1", "1"},
+		{"a2b1", "a2", "b1", "1"},
+
+		// Ensure seek will return on a tombstone if it is the latest key.
+		{"a3Xa1b1", "a3", "a3X", ""},
+		{"a3Xa1c2Xc1", "b1", "c2X", ""},
+
+		// Ensure seek does not return on a key shadowed by a tombstone.
+		{"a3Xa2a1b1", "a2", "b1", ""},
+		{"a3Xa2a1b1", "a2", "b1", "3"},
```
Comment (msbutler) [inline pkg/storage/backup_compaction_iterator_test.go:133]: this is a valid case`{"a3Xa2a1b1", "a2", "a2", "2"}`, yeah?

File: pkg/storage/backup_compaction_iterator_test.go (hunk 5 above)
Comment (msbutler) [reply pkg/storage/backup_compaction_iterator_test.go:142]: i'd add a few more cases where we seek to a tombstone. e.g. your seek key has timestamp later than the AsOF, and you end up on a tombstone.

**Review by msbutler — COMMENTED** [review]
Nice!

**Comment by kev-cao** [comment]
@msbutler Thanks for all the feedback! I've updated the tests/comments

**Review by sumeerbhola — COMMENTED** [review]
Could you provide some context for this PR, such as where this iterator will be used?

Does this need to implement `SimpleMVCCIterator`? Looking at the example of `ReadAsOfIterator`, it seems to me that all the callers only use the concrete type and call only `{SeekGE, Valid, NextKey, UnsafeKey, UnsafeValue}`. If that is correct, we should also remove unnecessary methods from `ReadAsOfIterator` (in a separate PR of course).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137529#-:-OELtSAU5VV0Khb8TOqm:b-ennn8s)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @kev-cao, and @msbutler)

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137529)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @kev-cao, and @msbutler)

___
*[`pkg/storage/backup_compaction_iterator.go` line 27 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137529#-OELybbX-XXZFF5Ducgf:-OELybbX-XXZFF5Ducgg:bgxdl6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e4959089993abe3b652f5f2e37e4b29be99416d/pkg/storage/backup_compaction_iterator.go#L27)):*
> ```Go
> 	asOf hlc.Timestamp
> 
> 	// valid tracks if the current key is valid
> ```

nit: missing period.
___
*[`pkg/storage/backup_compaction_iterator.go` line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137529#-OELyd8Z8eTRSkspmU4N:-OELyd8Z8eTRSkspmU4O:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e4959089993abe3b652f5f2e37e4b29be99416d/pkg/storage/backup_compaction_iterator.go#L30)):*
> ```Go
> 	valid bool
> 
> 	// err tracks if iterating to the current key returned an error
> ```

ditto
___
*[`pkg/storage/backup_compaction_iterator.go` line 63 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137529#-OELyYGeB7FwHxpV9MzQ:-OELyYGeB7FwHxpV9MzR:bgxdl6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e4959089993abe3b652f5f2e37e4b29be99416d/pkg/storage/backup_compaction_iterator.go#L63)):*
> ```Go
> 
> func (f *BackupCompactionIterator) SeekGE(originalKey MVCCKey) {
> 	// See ReadAsOfIterator comment for explanation of this
> ```

nit: missing period.
___
*[`pkg/storage/backup_compaction_iterator.go` line 76 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137529#-OELv1M1E2wTy4dsVqGg:-OELv1M1E2wTy4dsVqGh:bgxdl6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e4959089993abe3b652f5f2e37e4b29be99416d/pkg/storage/backup_compaction_iterator.go#L76)):*
> ```Go
> }
> 
> // advance moves past keys with timestamps later than f.asOf
> ```

nit: missing period.
___
*[`pkg/storage/backup_compaction_iterator.go` line 136 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137529#-OELvorQE_RQdgaUuGYI:-OELvorQE_RQdgaUuGYJ:b-m9m5g6) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e4959089993abe3b652f5f2e37e4b29be99416d/pkg/storage/backup_compaction_iterator.go#L136)):*
> ```Go
> 
> func (f *BackupCompactionIterator) assertInvariants() error {
> 	if err := assertSimpleMVCCIteratorInvariants(f); err != nil {
> ```

never mind my earlier comment -- I suppose reusing this function is why this implements `SimpleMVCCIterator`.
Those panics in the earlier methods don't trip up this invariant checking because all those `Range*` methods are only called when `HasPointAndRange` returns `hasRange=true`. As a nit, I am not sure that is properly documented as requirements in `SimpleMVCCIterator`, so consider strengthening the commentary in `SimpleMVCCIterator`.
___
*[`pkg/storage/backup_compaction_iterator.go` line 140 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137529#-OELwnGB7zUGJavsgNkq:-OELwnGB7zUGJavsgNkr:byg007r) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e4959089993abe3b652f5f2e37e4b29be99416d/pkg/storage/backup_compaction_iterator.go#L140)):*
> ```Go
> 	}
> 
> 	if f.asOf.IsEmpty() {
> ```

This assertion belongs in `NewBackupCompactionIterator`.
___
*[`pkg/storage/backup_compaction_iterator.go` line 146 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137529#-OELx7AQ6rUCtfW1iN_d:-OELx7AQ6rUCtfW1iN_e:b-v42x3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e4959089993abe3b652f5f2e37e4b29be99416d/pkg/storage/backup_compaction_iterator.go#L146)):*
> ```Go
> 	if ok, err := f.iter.Valid(); !ok || err != nil {
> 		errMsg := err.Error()
> 		return errors.AssertionFailedf("invalid underlying iter with err=%s", errMsg)
> ```

iterators can have errors because of underlying detected corruption. Such errors are not assertion failures. If the iterator did not detect an error (which happens via block checksumming etc.), we may still want to check application level (CRDB being the application) errors or certain invariants that should never be violated. Those should be wrapped in `AssertionFailedf`. If you look at `assertSimpleMVCCIteratorInvariants` that's roughly the pattern that is followed in examples like:

```
	// Keys can't be empty.
	if len(key.Key) == 0 {
		return errors.AssertionFailedf("valid iterator returned empty key")
	}
```

```
		value, err := iter.UnsafeValue()
		if err != nil {
			return err
		}
```

**Review by kev-cao — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137529)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @msbutler, and @sumeerbhola)

___
*[`pkg/storage/backup_compaction_iterator.go` line 136 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137529#-OELvorQE_RQdgaUuGYI:-OEQAa9F58gb6Izlwh3B:b-xuszxr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e4959089993abe3b652f5f2e37e4b29be99416d/pkg/storage/backup_compaction_iterator.go#L136)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

never mind my earlier comment -- I suppose reusing this function is why this implements `SimpleMVCCIterator`.
Those panics in the earlier methods don't trip up this invariant checking because all those `Range*` methods are only called when `HasPointAndRange` returns `hasRange=true`. As a nit, I am not sure that is properly documented as requirements in `SimpleMVCCIterator`, so consider strengthening the commentary in `SimpleMVCCIterator`.
</blockquote></details>

To provide some context on this work, feel free to check out the [prototype](https://github.com/cockroachdb/cockroach/pull/137040). Another reason for re-using `SimpleMVCCIterator` is because we will update the iterator in `MergedSST` to a generic type so that it can be used in both restores as well as compactions.

Regarding the requirement on `hasRange=true`, it looks like our existing underlying iterators simply return zero values if the range does not exist. @msbutler I did notice that `ReadAsOfIterator` also does the same, to preserve the behavior perhaps we should remove the panics.
___
*[`pkg/storage/backup_compaction_iterator.go` line 140 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137529#-OELwnGB7zUGJavsgNkq:-OEQAGTwAfgwJBiYnBHo:b-8xhxhm) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e4959089993abe3b652f5f2e37e4b29be99416d/pkg/storage/backup_compaction_iterator.go#L140)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This assertion belongs in `NewBackupCompactionIterator`.
</blockquote></details>

`NewBackupCompactionIterator` sets `asOf` to the `MaxTimestamp` if not provided (same behavior as the `ReadAsOfIterator`)
___
*[`pkg/storage/backup_compaction_iterator.go` line 146 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137529#-OELx7AQ6rUCtfW1iN_d:-OEQDjec1WI-TiTFOKMi:b-punda) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e4959089993abe3b652f5f2e37e4b29be99416d/pkg/storage/backup_compaction_iterator.go#L146)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

iterators can have errors because of underlying detected corruption. Such errors are not assertion failures. If the iterator did not detect an error (which happens via block checksumming etc.), we may still want to check application level (CRDB being the application) errors or certain invariants that should never be violated. Those should be wrapped in `AssertionFailedf`. If you look at `assertSimpleMVCCIteratorInvariants` that's roughly the pattern that is followed in examples like:

```
	// Keys can't be empty.
	if len(key.Key) == 0 {
		return errors.AssertionFailedf("valid iterator returned empty key")
	}
```

```
		value, err := iter.UnsafeValue()
		if err != nil {
			return err
		}
```
</blockquote></details>

From the looks of it, `assertSimpleMVCCIteratorInvariants` already wraps errors that should be considered assertion errors as assertion errors, so I suppose just returning the error as is is sufficient.

**Review by kev-cao — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137529)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @msbutler, and @sumeerbhola)

___
*[`pkg/storage/backup_compaction_iterator.go` line 146 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137529#-OELx7AQ6rUCtfW1iN_d:-OEQETAa35jFCe14WYJp:b-4aciwz) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e4959089993abe3b652f5f2e37e4b29be99416d/pkg/storage/backup_compaction_iterator.go#L146)):*
<details><summary><i>Previously, kev-cao (Kevin Cao) wrote…</i></summary><blockquote>

From the looks of it, `assertSimpleMVCCIteratorInvariants` already wraps errors that should be considered assertion errors as assertion errors, so I suppose just returning the error as is is sufficient.
</blockquote></details>

Oh whoops was reading the wrong section — in this case, `f.iter.Valid()` is being checked if `f.valid` is true, I think that's a fine assertion to make. Also the `assertSimpleMVCCIteratorInvariants` call at the top handles those assertions. I'm not quite sure what you change you are proposing here, could you elaborate?

**Comment by msbutler** [comment]
@sumeerbhola for more context, checkout this design doc https://docs.google.com/document/d/1ZslwAxnr46VvrwOL-P_rAUnWPo2ztXl8F2SkoB3VpK4/edit?tab=t.0#heading=h.9ndo02nsmu

**Review by sumeerbhola — COMMENTED** [review]
Regarding the context, why are there no range keys (MVCC tombstones) handled here? Is this related to the idea that once a range tombstone is applied, no point keys will ever be written on top (an invariant maintained by a higher layer), so the `BackupCompactionIterator.iter` is already configured to apply the effect of the range tombstone but not surface it?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137529#-:-OEU80jT3km3i7I1uVBa:b-dlybhs)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @kev-cao, and @msbutler)

**Comment by msbutler** [comment]
> Regarding the context, why are there no range keys (MVCC tombstones) handled here? 

At least for the MVP of this backup compaction feature, we are assuming that rangekeys will not get written to the key space that we need to back up. Jeff is currently exploring the performance implications of rolling back imports with point tombstones instead of range tombstones, for example.

**Review by sumeerbhola — DISMISSED** [review]
> Reviewed 1 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137529)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @kev-cao, and @msbutler)

___
*[`pkg/storage/backup_compaction_iterator.go` line 136 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137529#-OELvorQE_RQdgaUuGYI:-OEX7RfG1o4WEQo-0qsj:b-j5va25) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e4959089993abe3b652f5f2e37e4b29be99416d/pkg/storage/backup_compaction_iterator.go#L136)):*
<details><summary><i>Previously, kev-cao (Kevin Cao) wrote…</i></summary><blockquote>

To provide some context on this work, feel free to check out the [prototype](https://github.com/cockroachdb/cockroach/pull/137040). Another reason for re-using `SimpleMVCCIterator` is because we will update the iterator in `MergedSST` to a generic type so that it can be used in both restores as well as compactions.

Regarding the requirement on `hasRange=true`, it looks like our existing underlying iterators simply return zero values if the range does not exist. @msbutler I did notice that `ReadAsOfIterator` also does the same, to preserve the behavior perhaps we should remove the panics.
</blockquote></details>

The change to remove the panics is good enough -- no need to update the `SimpleMVCCIterator` interface.
___
*[`pkg/storage/backup_compaction_iterator.go` line 140 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137529#-OELwnGB7zUGJavsgNkq:-OEX5WXpBU6oc-YHfa4d:b-6914th) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e4959089993abe3b652f5f2e37e4b29be99416d/pkg/storage/backup_compaction_iterator.go#L140)):*
<details><summary><i>Previously, kev-cao (Kevin Cao) wrote…</i></summary><blockquote>

`NewBackupCompactionIterator` sets `asOf` to the `MaxTimestamp` if not provided (same behavior as the `ReadAsOfIterator`)
</blockquote></details>

Sure, but I am not clear on why an assertion about an immutable field of `f` belongs in a method that will be called frequently. If the concern is that someone will not use `NewBackupCompactionIterator` and just initialize the fields, perhaps hide this in a different package (though I don't think we need to be that paranoid).
___
*[`pkg/storage/backup_compaction_iterator.go` line 146 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137529#-OELx7AQ6rUCtfW1iN_d:-OEX6bnX6XxdNysfsRsO:b-he4gdo) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e4959089993abe3b652f5f2e37e4b29be99416d/pkg/storage/backup_compaction_iterator.go#L146)):*
> in this case, `f.iter.Valid()` is being checked if `f.valid` is true, I think that's a fine assertion to make.

I missed that. Can you add a code comment stating this here, and add something like the following at the top of the method:
```
// REQUIRES: f.valid
func (f *BackupCompactionIterator) assertInvariants() error {
```

**Review by msbutler — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137529)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @kev-cao, and @sumeerbhola)

___
*[`pkg/storage/backup_compaction_iterator.go` line 140 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137529#-OELwnGB7zUGJavsgNkq:-OE_QK0i5k3DHRIupGbp:b-ed7ezm) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e4959089993abe3b652f5f2e37e4b29be99416d/pkg/storage/backup_compaction_iterator.go#L140)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Sure, but I am not clear on why an assertion about an immutable field of `f` belongs in a method that will be called frequently. If the concern is that someone will not use `NewBackupCompactionIterator` and just initialize the fields, perhaps hide this in a different package (though I don't think we need to be that paranoid).
</blockquote></details>

+1 to Sumeer's point. It seems like NewBackupCompactionIterator should simply return an error if the user passes a nil asOf.

**Review by msbutler — APPROVED** [review]

**Comment by kev-cao** [comment]
TFTR!

bors r=msbutler
